### ISSUE TITLE
Performance issue with > 1000 markers and cluster

### ISSUE THREAD

**Issue by jbjuin** [thedirtyfew/dash-leaflet#24]
Hi, thanks a lot for this awesome component ! Great work !

I have few questions:

- First thing is that the "click_lat_lng" on Marker seems to be absent of 0.0.17. You mention this addition in multiple issues, referencing 0.0.4. Have those 2 tags diverged somewhere ?
- Since click_lat_lng is present on Circle, we used this instead of Marker. So that part works... but we have huge performance problems with > 1000 Circles... clicking on one item takes few seconds before the callback is called and update a "log" with clicked marker id... did you experience this as well ?

I don't have exeperience with react so I may not be able to help a lot here.

Anyway keep up the good work on this !

**Comment by emilhe** [comment]
I have started working on events for the marker cluster component. One of the properties i had in mind was `click_marker_id`, which holds the id of the latest marker clicked. I just tried the previously posted example with the new syntax, i.e.

```
import json
import random
import dash
import dash_html_components as html
import dash_leaflet as dl
from dash.dependencies import Output, Input

n = 10000
lats = [random.randrange(44000, 48000) / 1000.0 for i in range(n)]
lons = [random.randrange(0, 4000) / 1000.0 for i in range(n)]
markers = [dl.Marker(id=str(i), position=(lats[i], lons[i])) for i in range(n)]
cluster = dl.MarkerClusterGroup(children=markers, id="cluster")

app = dash.Dash(external_stylesheets=['https://codepen.io/chriddyp/pen/bWLwgP.css'], prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map(children=[dl.TileLayer(url="https://a.tile.openstreetmap.org/{z}/{x}/{y}.png"), cluster],
           style={'width': "100%", 'height': "400px"}, center=[46, 2], zoom=7, id="map", preferCanvas=True), html.P(id="log")],
    style={"position": "relative", 'width': '1000px', 'height': '300px'})


@app.callback(Output("log", "children"), [Input("cluster", "click_marker_id")])
def marker_click(marker_id):
    return marker_id

if __name__ == '__main__':
    app.run_server(debug=False)
```

and the performance it somewhat better, though still not blazing fast (for 10.000 markers, the previous code took 1300 ms on my desktop vs 300 ms for the new implementation). If you want to try it out, you can get the code like this,

`pip install dash-leaflet==0.0.18rc2`

Please note that the syntax (e.g. the name; should it be `click_marker_id`? or `marker_click_id` or something else?) is still subject to change.

**Comment by jbjuin** [comment]
Hi, thanks for this update !

I will try this as soon as I can. 

On our side we started to implement a "simpler" solution with a very simple wrapper around leaflet without wrapping all items into react components and integrate prunecluster. 

We reach ~100ms reaction on click on a marker with 400.000 clustered points.

The pros of your proposed component is that everything is customizable... in our solution we still have a lot of work to bind props... so we may switch back to this one if the performance are enough.

Thanks again for that project. I come back to comment and provide feedback as soon as I did my tests.

**Comment by emilhe** [comment]
Wow that is impressive! Could you link to the code? I am curious why the performance difference is so big!

**Comment by jbjuin** [comment]
OK. I don't think that clustering is the problem, is it ? I think I tested dash-leaflet without clustering and there was still a perceivable delay on marker click with 10.000 markers...

**Comment by emilhe** [comment]
@jbjuin - Yes and no. Without clustering, **all** the markers must be created, which slows down the map. With clusters, only a few markers (the cluster "bubbles") are drawn on the map, which is much faster. Hence, from this point of view, the clustering actually speeds up the map. However, the clustering also introduces some additional calculations, which might slow down the map. For the markercluster package that i am using, this seems to become a problem for more than a few 1000 markers. The prunecluster on the other hand seems to be able to handle a lot more markers, i guess due to a more efficient clustering algorithm.

**Comment by zeluspudding** [comment]
Glad to find this thread - I am having similar performance issues when clustering 26,000+ points. Except, I'm not sure if I should continue in this thread or start a new one so just let me know if I should start a new one.

I am using the `dl.MarkerClusterGroup`. 

I noticed that there is a [chunkedLoading](https://github.com/Leaflet/Leaflet.markercluster#chunked-addlayers-options) option in the original leaflet.js so tried enabling that option as follows:
![image](https://user-images.githubusercontent.com/8431341/86975755-9dee6180-c13e-11ea-905c-8c82ac0e18de.png)

Sadly that didn't seem to translate into any performance improvements. I did a little search/`CTRL F`ing through the dash-leaflet [source code](https://github.com/thedirtyfew/dash-leaflet/tree/master/src/lib) to see of any mention of a `chunk` option but, alas, did not see any. So I either missed it or it's not currently available. Did this option make it into your port?

I also tried setting the `preferCanvas` option as recommended above but that didn't seem to speed anything up :/ Either way it takes about 25 seconds to load... and also appears to drop about 10,000 points. Can see it [drag here](https://dash.scious.io/)

**Comment by zeluspudding** [comment]
Man, I really enjoy the UI of leaflets cluster map in your Dash port - it's so nice for example how it consolidates/arranges many pins associated with the same point

![image](https://user-images.githubusercontent.com/8431341/86999539-6b615a80-c178-11ea-9c5e-86d912e2eb9c.png)

 if there is anyway to improve the performance to get 10s of thousands of these in cluster, I am all ears. Glad to help as possible too :)

**Comment by zeluspudding** [comment]
Note - there is activity on this topic posted on the Dash forum [here](https://community.plotly.com/t/how-to-use-mapbox-clusters/10774/4?u=agosta)

**Comment by emilhe** [comment]
I haven't made any efforts to explicitly support `chunkedLoading` (which is why you cannot find it in the source code), but since the `options` are passed directly to the JavaScript layer, i don't see why it shouldn't work. 

That being said, i don't think that `chunkedLoading` is the solution. To me it seems to just postpone the issue slightly. In order to really improve the handling of many markers, i think the proper solution would be to change the underlying clustering algorithm to use supercluster (like mapbox) or the algorithm used by PruneCluster.

**Comment by zeluspudding** [comment]
@emilhe , thank you for your response! :)

### On `chunkedLoading`
Yea, I'm not sure if this option is getting engaged or, like you said - if it is, it's not the solution so not having a significant impact. That's my perception after running network profiles under these conditions:

__15ish seconds before significant visible dom rendering__
> "chunkedLoading": True,
> preferCanvas=True,


__10ish seconds before significant visible dom rendering__
> "chunkedLoading": True,
> #preferCanvas=True,  ##(so no canvas)


__15ish seconds before significant visible dom rendering__
> #"chunkedLoading": True, ##(so no chunked loading)
> preferCanvas=True,

Judging by those numbers - it's a wash.

### Support of slow clustering algorithm as culprit
In every profile I observed a 10 second gap between initial loading of resources and when leaflet loads its map tiles:
![image](https://user-images.githubusercontent.com/8431341/87058785-1bad7e00-c1ce-11ea-8bc2-a422caa8d93b.png)

... to me this seems to support your idea that the clustering algorithm could be responsible for the delay we're seeing.

I tried to look into what the CPU was doing during this time. Not sure I used the right tool for this but here's what Firefox had to report on the matter:

 - there were about 18 seconds of rendering activity taking place, much of which seemed a play between `dash renderer` and `dash_leaflet`.
 - there was 11 seconds of idle time which I don't understand.
![image](https://user-images.githubusercontent.com/8431341/87059379-db023480-c1ce-11ea-8cc3-029ac0a1d5ae.png)

### JSON payload optimization
As seen above, the JSON payload the browser received is over 6MB. 


### Missing markers / json payload optimization
None of this is able to explain why over 10k markers are missing. Will have to dig into this more. ### JSON payload optimization
The JSON payload the browser gets is over 6MB.

**Comment by zeluspudding** [comment]
### Circling back on these missing markers
Turns out I'm actually missing well over 20,000 markers. At nearly 27k markers in my cluster:

![image](https://user-images.githubusercontent.com/8431341/87069089-3ab30c80-c1dc-11ea-8aaf-066f481422f5.png)

only about 2.7k get mapped:

![image](https://user-images.githubusercontent.com/8431341/87069192-63d39d00-c1dc-11ea-9ebb-bbe8b47ba496.png)

I doesn't appear that this is because my lat-lon pairs are out of bounds... but perhaps I am wrong?
![image](https://user-images.githubusercontent.com/8431341/87068967-050e2380-c1dc-11ea-8897-98eb9cc779ed.png)

**Comment by emilhe** [comment]
It would definitely be nice to reduce the JSON payload size. However, as i understand, the kind of optimizations that you are suggesting (e.g. renaming children) would need to be implemented in Dash itself. From a dash component (as Dash Leaflet is), you can only change the object structure - not how the structure is serialized to JSON. 

That being said, it would be possible to redesign the cluster component in such a way that the JSON payload is reduced. As an example, say rather that than a list of children that the cluster component would take as arguments list of coordinates along with a single marker (which would be placed at all positions). This structure would reduce the JSON payload **a lot** (all the overhead that you point out would essentially be gone), but at the cost of flexibility.

If it turns out that the JSON payload is the main issue, i will look further into how it can be reduced without reducing flexibility too much. However, for now i still think that there are other issues (the clustering algorithm, and possible also the dash rendering) that might be more prominent. You benchmarks look really cool - please let me know if you come up with more definitive answers on what the actual performance bottlenecks are.

**Comment by zeluspudding** [comment]
Hey! Thanks for that message :) 

Yes, as you indicated, Dash rendering seems to be a bottleneck.. or at least, it is being held up by the 7 second average load time of async-plotlyjs.js as [articulated here.](https://community.plotly.com/t/smaller-version-of-async-plotlyjs-js-its-so-big/42247)

**Comment by zeluspudding** [comment]
Perchance any thoughts on the dropped markers?

**Comment by zeluspudding** [comment]
Little update here - and maybe I should file this into a new issue:
When zoomed out, the leaflet cluster indicates that there are no more than 3k markers in the main cluster. But when zoomed in it shows several clusters that have over 4.5k markers.

![](https://i.imgur.com/EgMmQPL.gif)

So it looks like the data is actually in the map, it's just not getting shown correctly. Any ideas why that may be?

**Comment by emilhe** [comment]
I have never seen anything like the dropped markers, but my guess would be that you are seeing a bug in the underlying JS library. Are you able to share the data for testing?

**Comment by zeluspudding** [comment]
Hey @emilhe, thank you for your message :) [Here is a link to a self contained dash app example](https://drive.google.com/drive/folders/13jA8RjPg-_4J_0vOXWZsZ7B1wgnw1Yod?usp=sharing) that includes the data (note that running the app via `debug_mode=True` will cause it to take a really long time to load which is why I have it running in `debug_mode=False`. 

#### JSON payload
> If it turns out that the JSON payload is the main issue, i will look further into how it can be reduced without reducing flexibility too much. However, for now i still think that there are other issues (the clustering algorithm, and possible also the dash rendering) that might be more prominent.

Having tinkered quite a bit over the last two days, I am thinking you are right about `dash_renderer` being an issue, except maybe with a twist - it's the combination of  `dash_renderer` having to parse the large JSON payload structure that `dash_leaflet` is sending. 

As seen below, there is an average 1.5 seconds the client is waiting to receive the Dash Layout payload (which includes the Leaflet payload of 20k Markers). I imagine much of this wait is on Dash server bundling the Markers into a Dash Layout and then compressing this for delivery client side.

![image](https://user-images.githubusercontent.com/8431341/87230875-3a4f7880-c378-11ea-9191-17e6f7164d01.png)

Once client side, the 20k Markers need to get added to the DOM so I think that's why we see a pause in network activity between seconds 4 and 8.5 above. Part of this time - I'm assuming - is decompressing the payload but most of it is indeed spent adding Marker elements to the DOM, as seen from the Chrome Dev Tools Performance tab below (which has a different timeline only because Chrome Dev Tools don't allow you to monitor Network and CPU performance at the same time and I had to run this benchmark against a separate page reload).

![image](https://user-images.githubusercontent.com/8431341/87231142-39b7e180-c37a-11ea-8941-c339bd494213.png)

The big yellow block in the image above is the time tied up with javascript adding markers to DOM. ... Looking at memory allocation confirms this.

So if you think there is a way to reformat the Marker's Payload to have a smaller memory footprint, I am all ears and hands on deck to help with that. I don't have a dash components development setup just yet but if I can be of assistance by digging into the code that way and/or testing on my side - I'm glad to do that :) Let me know how I can help

**Comment by emilhe** [comment]
I just did some quick testing. I think that the markers are actually _not_ missing. In e.g. the image below, it might look like there are only "2553" markers in the blob,

![cluster](https://user-images.githubusercontent.com/1623542/87234811-1bcb9a80-c3d5-11ea-9bd8-9c9cbc236c56.png)

But if you look closely, a "5" is placed below the bubble, i.e. the real number is probably "25535", but with the last digit printed on the next line. I guess you can avoid the splitting of numbers across lines by adjusting some CSS.

I just tried out the performance optimization i suggested in the previous post, and it speeds up your example **a lot**. Hence i think that you are right that the large JSON payload is responsible for a significant part of the performance issues. I will dig more into it tomorrow.

**Comment by zeluspudding** [comment]
This is so cool! Will stand by excitedly :)

As for the carriage return on the fifth significant figure, I'm embarrassed I didn't catch that. Looking into the CSS to fix that. Thank you so much for your help :) Also, does this project by any chance have a way of receiving donations? I am interested in supporting this project :)

**Comment by emilhe** [comment]
Changing the structure of the cluster component to accept a list of coordinates (rather than a list of markers), i was able to reduce the rendering time of your example by around 50% (from ~3s to ~1.5s on my laptop). The improvement seems to be mainly related to speed up of the dash renderer.

After doing a little digging into the function calls, it seems that most of the remaining time is spent constructing the Marker objects, first in React and then in Leaflet. It also explains why you see a speed up when omitting Tooltip object; when they are present, event more objects must be constructed. 

Hence, for now, i would say that the key to improve (loading) performance would be to (somehow) delay the construction of markers until they are actually needed. Implementing this kind of behavior from scratch would require some work though. Hence i am not sure it would be worth it compared to just porting e.g. Mapbox GL, which has these features built in.

**Comment by emilhe** [comment]
I have just made a quick and dirty implementation of a `SuperCluster ` component to get an idea about potential performance improvements. And it is **fast**. Like _really_ fast. Here is your example with ~ 30k points,

![example](https://user-images.githubusercontent.com/1623542/87299364-28064380-c50c-11ea-956b-b27bcb385a6a.gif)

On my laptop, the app renders in about 200 ms. In comparison a blank Dash app takes around 50 ms to load, i.e. the map load in about 150 ms!

**Comment by zeluspudding** [comment]
Waaaaat!? This is amazing! How is that even possible :)

**Comment by emilhe** [comment]
I am running the example locally on my laptop, so the JSON file is available on disk. If you would have to download it, the data transfer (in the current form, around 3 MB) would of course take some time. But unless you connection is _very_ slow, download 3 MBs shouldn't take that long.

**Comment by emilhe** [comment]
Only was only referring to the _data loading_ taking place in the JS layer. The component is created from Dash, i.e. it looks like this,

```
import dash
import dash_html_components as html
import dash_leaflet as dl

app = dash.Dash(prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.SuperCluster(data='assets/myfile.geojson')],  center=(42.54, -78.05), zoom=5,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
])

if __name__ == '__main__':
    app.run_server(debug=False)
```

I also tried passing the data directly. It adds a slight delay (seem to be around ~ 200 ms), but it is not that much.

**Comment by zeluspudding** [comment]
Oh wow!! Very very cool :)

Like I've said before, I'm glad to help test, measure and debug however needed. Just let me know. This is so cool!

**Comment by emilhe** [comment]
I just pushed a dev version (0.0.21rc1) with the new component. If you install it, you should be able to run the following example,

```
import dash
import dash_html_components as html
import dash_leaflet as dl
import json
from dash.dependencies import Output, Input

app = dash.Dash(prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.SuperCluster(data='assets/myfile.geojson', id="supercluster")],  center=(42.54, -78.05), zoom=5,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
           html.Div(id="n_clicks"),
           html.Div(id="marker_click")
])

@app.callback(Output("n_clicks", "children"), [Input("supercluster", "n_clicks")])
def func(n_clicks):
    return n_clicks

@app.callback(Output("marker_click", "children"), [Input("supercluster", "marker_click")])
def func(marker_click):
    return json.dumps(marker_click)

if __name__ == '__main__':
    app.run_server()
```

Note that the data must now be supplied as GeoJSON. I have attached your data converted to geojson af an example (just rename it to .geojson).

[myfile.txt](https://github.com/thedirtyfew/dash-leaflet/files/4913018/myfile.txt)

**Comment by zeluspudding** [comment]
This is soo cool! Will play with this this evening :)

**Comment by zeluspudding** [comment]
Okay, I was able to install dev version `0.0.21rc2` and ran it. Got the following error:
```
TypeError: The `dash_leaflet.SuperCluster` component (version 0.0.21rc1) received an unexpected keyword
argument: `url` 
Allowed arguments: data, id, marker_click, n_clicks, options, zoomToBoundsOnClick
```
so I changed `url` to data (even though I knew it was still going to throw an error) and then got the following error
```
TypeError: The `dash_leaflet.SuperCluster` component (version 0.0.21rc1) received an unexpected keyword
argument: `format` 
Allowed arguments: data, id, marker_click, n_clicks, options, zoomToBoundsOnClick
```

Thus it really doesn't like the line in `dl.SuperCluster(url='assets/leaflet_50k.pbf', format="geobuf", options={"radius": 100})],` in:
```
dl.Map([dl.TileLayer(),
	dl.SuperCluster(url='assets/leaflet_50k.pbf', format="geobuf", options={"radius": 100})],
	center=(-37.79, 175.27),
	zoom=13,
	style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"})
),
```

Can I feed a `geopandas` dataframe (as geojson) to this `data` argument?

**Comment by emilhe** [comment]
From you error it seems you installed the wrong version?

**Comment by zeluspudding** [comment]
Oh wow! I just upgraded to `0.0.21rc2`. The difference is night and day :) Very exciting! Am now playing with injecting my data from Dash (rather than file).

**Comment by emilhe** [comment]
Great! Ideally, how would you inject the data? Lists? Or maybe a data frame?

**Comment by zeluspudding** [comment]
Ideally as a dataframe since that's the form the data is originally in. I
suppose that geopandas could be used as well but it looks like one
could achieve
the same effect by only importing `geojson`
<https://gis.stackexchange.com/a/221040/139332> which I presume is lighter
than `geopandas`. But if it absolutely had to be converted to a list, then
I could always do `dataframe.to_list()`.. That just adds a little overhead
for the conversion but is what I was doing earlier to generate markers for
the prior/non-supercluster version of dash-leaflet.

**Comment by emilhe** [comment]
By a list, i was thinking along the lines of a list of records. Hence, the syntax wold be something like,

```
markers = [
    dict(lat=55, lon=10, tooltip="I am a tooltip"),  # marker with tooltip
    dict(lat=56, lon=10, popup="I am a popup"),  # marker with popup
    dict(lat=55, lon=11, markerOptions=dict(opacity=0.5)),  # marker with marker options, in this case a change in opacity
    dict(lat=56, lon=12, markerOptions=dict(icon=dict(iconUrl="assets/icon_plane.png"))),  # marker custom icons options, in this case a plane as icon
]
```

If you had a dataframe with the data,

```
data = [
    [55, 10, "I am a tooltip", None, None, None],
    [56, 10, None, "I am a popup", None, None],
    [55, 11, None, None, "0.5", None],
    [56, 11, None, None, None, "assets/icon_plane.png"]
]
df = pd.DataFrame(data=data, columns=["lat", "lon", "tooltip", "popup", "opacity", "icon"])
```

the conversion would be something like,

```
def row_to_marker(row):
    marker = {key: row[key] for key in df.columns if row[key]}
    # Handle non-direct mappings.
    if "opacity" in marker:
        marker["markerOptions"] = dict(opacity=0.5)
    if "icon" in marker:
        marker["markerOptions"] =dict(icon=dict(iconUrl="assets/icon_plane.png"))
    return marker
    
markers = df.apply(row_to_marker, axis=1)
```

Since a lot of marker customization options are dictionaries, i don't see any other options that delegating the dataframe-to-maker conversion to the user. If you have suggestions for a better syntax, let me know.

**Comment by zeluspudding** [comment]
That makes sense! Although I'm curious if that format brings us back to the original issue of the payload getting large (~`6.6MB`). Then again in this case I'm not sure how much leaner it could be made aside from minifying key names (i.e. from `opacity` to `op`)

**Comment by emilhe** [comment]
The format that we that discussing now (e.g. the `markers` structure above) is only for use as a wrapper in Python; the actual data will be transferred as GeoJSON. For the structure above, the conversion would look something like this,

**Comment by zeluspudding** [comment]
Gotcha!! That makes perfect sense - thank you for explaining this :)

**Comment by emilhe** [comment]
I have now (0.0.21rc3) implemented some util methods to ease the construction of markers from python. I have also implemented tooltips, popups, marker options and custom icons. Here is an updated example,

```
import dash
import dash_html_components as html
import dash_leaflet as dl
import dash_leaflet.express as dlx
import geobuf
import base64
from numpy import random

# Create some simple markers.
markers = [dict(lat=56 + random.rand(), lon=10 + random.rand()) for i in range(10000)]
# Create a few customized markers.
markers += [
    dict(lat=55, lon=9, tooltip="I am a tooltip"),  # marker with tooltip
    dict(lat=55, lon=10, popup="I am a popup"),  # marker with popup
    dict(lat=55, lon=11, markerOptions=dict(opacity=0.5)),  # marker with options
    dict(lat=55, lon=12, markerOptions=dict(icon=dict(iconUrl="assets/icon_plane.png"))),  # marker with custom icon
]
# Convert markers to geobuf (smaller payload than raw geojson).
geojson = dlx.markers_to_geojson(markers)
pbf = base64.b64encode(geobuf.encode(geojson)).decode()
# Create example app.
app = dash.Dash()
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.SuperCluster(data=pbf, format="geobuf", options={"radius": 100})], 
    center=(56, 10), zoom=7, style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"})
])

if __name__ == '__main__':
    app.run_server(debug=False)
```

There are still a lot of things that could be added,

* Customization of the cluster icon (currently not possible)
* Fancy animations (e.g. spiderfy)
* Support of more formats, e.g. gzip
* Support of children in marker popup/tooltips
* ....

but i feel that the component is getting close to be "good enough" for release in a first version.

**Comment by zeluspudding** [comment]
This is so amazing @emilhe ! I will be testing this tonight! 

What are you predicting it would take to implement spiderfy? I ask because when zoomed in I've currently got 1000 markers that look like they're only one marker because they're all mapped onto the same spot :/ I think this is already the best cluster mapping package for dash but an addition of 'spiderfy' pushes that up by far. Like I said before, if there is any way I can help - including through donation support - I'd love to do that :) Will get back to you with test results tonight :)

**Comment by zeluspudding** [comment]
Holy cow @emilhe this is sooo amazing!!! Sorry I'm just getting back to you now- finally got time to implement. This clustering improvement is blazing fast! Have it hosted at [dash.scious.io](https://dash.scious.io/) and it's a night and day difference from the original loading speeds. 

By the way, that was very clever what you did with `geobuf` :) I didn't even know that was a thing!

Are you sure there's no way I can support this project with a donation? I would love to help out however I can :)

**Comment by emilhe** [comment]
Great to hear that! You pages looks cool - and the map does indeed load fast ;). I have now finally found the time to wrap of the first version of the component. Compared to the previous version, i have fixed a few bugs and added the following features

* Customization of the cluster classes (which css classes to apply and when)
* Spiderfy functionality (!)
* Events (both click and hoover)

I have also added an [example in the documentation](http://dash-leaflet.herokuapp.com/#super_cluster). And last, but not least, i have added a donate button in the readme :)

**Comment by ELGUMOS1** [comment]
Hey guys I know it is not the place to ask this question but it is kind of related and I have been trying some patters without success.
I am struggling to get the marker id from a LayerGroup by clicking on it, the thing is that the markers are not in a global variable to be accessed using:

@app.callback(Output("log", "children"), [Input(marker, "click_lat_lng") for marker in markers])

def marker_click(*args):
    if not any(list(args)):
        return ""
    marker_id = dash.callback_context.triggered[0]['prop_id'].split(".")[0]
    return "You clicked marker {}".format(marker_id)

but they are generated dynamically via callback output in a list using the info from some dropdowns something like this:

dl.Map([dl.TileLayer(), dl.LayerGroup(id="layer")],zoom=12, center=(-33.4569397,-70.6482697),
                       id="map", style={'width': '100%', 'height': '100vh', 'margin': "auto",})

@app.callback([Output("layer", "children"),Output('download-link', 'href')],
              [Input('area_name', 'value'),
               Input('score_level', 'value'),
               Input('column-selector','value'),])
def update_figure(chosen_area,chosen_level,column):
    return my_tab3.update_figure(chosen_area,chosen_level,column,)

Does anyone know the correct pattern to access the marker id by clicking on it if the marker list is generated dynamically via callback?
Thank you in advance.

**Comment by zeluspudding** [comment]
Hey @emilhe  I just donated!! :) Just wanted to say thank you again for this super cool library!

One question I did have, having just upgraded to the latest release `0.0.21`, is there an option I am missing to keep the spiderfied lines drawn while panning/zooming in? The lines currently disappear at max zoom and make it seem like each marker is associated with different places. 

![](https://i.imgur.com/d1bJoHs.gif)

**Comment by emilhe** [comment]
@zeluspudding Thanks!! This issue(s) you are seeing probably related to three different point,

* There is a bug in (0.0.21) causing the markers to stay in the spiderfied positions. It should be fixed in (0.0.22).

* In my demo, the `maxZoom` of the `SuperCluster` (defaults to 16) is not the same as the `maxZoom` of the `TileLayer` (which is 18 for the default one), which is why the clustering stops at the highest zoom levels. In most cases, you would probably want these two zoom levels to be the same, so i have implemented some auto detection logic (in 0.0.22), which should work if you add the  `TileLayer` before the `SuperCluster`.

* I implemented the spiderfy logic with the intention to spiderfy a cluster, when you click it, if its children cannot be resolved at the max zoom level. When you change the view port, the spider is "collapsed" back to the cluster state. Would you prefer a different behavior?

**Comment by emilhe** [comment]
@ELGUMOS1 For this kind of use case, you would probably need to use [Pattern-Matching Callbacks](https://dash.plotly.com/pattern-matching-callbacks). It quite a bit off topic though, so i won't post an example here. The question is better suited for the Plotly Dash forum or Stackoverflow.

**Comment by zeluspudding** [comment]
Of course, you're very welcome!! Thank you! :)

> - In my demo, the maxZoom of the SuperCluster (defaults to 16) is not the same as the maxZoom of the TileLayer (which is 18 for the default one), which is why the clustering stops at the highest zoom levels. In most cases, you would probably want these two zoom levels to be the same, so i have implemented some auto detection logic (in 0.0.22), which should work if you add the TileLayer before the SuperCluster.

I agree - in most cases folks will probably want the maxZoom of TileLayer to be the same as SuperCluster's

> - I implemented the spiderfy logic with the intention to spiderfy a cluster, when you click it, if its children cannot be resolved at the max zoom level. When you change the view port, the spider is "collapsed" back to the cluster state. Would you prefer a different behavior?

I think that's exact the desired behavior :) Although perhaps with one tiny modification - rather than "collapse" the spider back to the cluster state on `viewport change` maybe it could be collapsed on `zoom change` (which in practice could only be a zoom out since, at this state, the map would be at full zoom). That would match the behavior of your standard leaflet dash component. But this seems like a minor thing. This component is already very useful!! :)

**Comment by ELGUMOS1** [comment]
> I have started working on events for the marker cluster component. One of the properties i had in mind was `click_marker_id`, which holds the id of the latest marker clicked. I just tried the previously posted example with the new syntax, i.e.
> 
> ```
> import json
> import random
> import dash
> import dash_html_components as html
> import dash_leaflet as dl
> from dash.dependencies import Output, Input
> 
> n = 10000
> lats = [random.randrange(44000, 48000) / 1000.0 for i in range(n)]
> lons = [random.randrange(0, 4000) / 1000.0 for i in range(n)]
> markers = [dl.Marker(id=str(i), position=(lats[i], lons[i])) for i in range(n)]
> cluster = dl.MarkerClusterGroup(children=markers, id="cluster")
> 
> app = dash.Dash(external_stylesheets=['https://codepen.io/chriddyp/pen/bWLwgP.css'], prevent_initial_callbacks=True)
> app.layout = html.Div([
>     dl.Map(children=[dl.TileLayer(url="https://a.tile.openstreetmap.org/{z}/{x}/{y}.png"), cluster],
>            style={'width': "100%", 'height': "400px"}, center=[46, 2], zoom=7, id="map", preferCanvas=True), html.P(id="log")],
>     style={"position": "relative", 'width': '1000px', 'height': '300px'})
> 
> 
> @app.callback(Output("log", "children"), [Input("cluster", "click_marker_id")])
> def marker_click(marker_id):
>     return marker_id
> 
> if __name__ == '__main__':
>     app.run_server(debug=False)
> ```
> 
> and the performance it somewhat better, though still not blazing fast (for 10.000 markers, the previous code took 1300 ms on my desktop vs 300 ms for the new implementation). If you want to try it out, you can get the code like this,
> 
> `pip install dash-leaflet==0.0.18rc2`
> 
> Please note that the syntax (e.g. the name; should it be `click_marker_id`? or `marker_click_id` or something else?) is still subject to change.

D

> I have started working on events for the marker cluster component. One of the properties i had in mind was `click_marker_id`, which holds the id of the latest marker clicked. I just tried the previously posted example with the new syntax, i.e.
> 
> ```
> import json
> import random
> import dash
> import dash_html_components as html
> import dash_leaflet as dl
> from dash.dependencies import Output, Input
> 
> n = 10000
> lats = [random.randrange(44000, 48000) / 1000.0 for i in range(n)]
> lons = [random.randrange(0, 4000) / 1000.0 for i in range(n)]
> markers = [dl.Marker(id=str(i), position=(lats[i], lons[i])) for i in range(n)]
> cluster = dl.MarkerClusterGroup(children=markers, id="cluster")
> 
> app = dash.Dash(external_stylesheets=['https://codepen.io/chriddyp/pen/bWLwgP.css'], prevent_initial_callbacks=True)
> app.layout = html.Div([
>     dl.Map(children=[dl.TileLayer(url="https://a.tile.openstreetmap.org/{z}/{x}/{y}.png"), cluster],
>            style={'width': "100%", 'height': "400px"}, center=[46, 2], zoom=7, id="map", preferCanvas=True), html.P(id="log")],
>     style={"position": "relative", 'width': '1000px', 'height': '300px'})
> 
> 
> @app.callback(Output("log", "children"), [Input("cluster", "click_marker_id")])
> def marker_click(marker_id):
>     return marker_id
> 
> if __name__ == '__main__':
>     app.run_server(debug=False)
> ```
> 
> and the performance it somewhat better, though still not blazing fast (for 10.000 markers, the previous code took 1300 ms on my desktop vs 300 ms for the new implementation). If you want to try it out, you can get the code like this,
> 
> `pip install dash-leaflet==0.0.18rc2`
> 
> Please note that the syntax (e.g. the name; should it be `click_marker_id`? or `marker_click_id` or something else?) is still subject to change.

I tried this and I did not get the marker id in the log or printed?
Any update, is kinda what I am looking for.

**Comment by emilhe** [comment]
@ELGUMOS1 In the new example in the documentation, you can see how to capture both click and hover events,

http://dash-leaflet.herokuapp.com/#super_cluster

**Comment by ELGUMOS1** [comment]
> @ELGUMOS1 In the new example in the documentation, you can see how to capture both click and hover events,
> 
> http://dash-leaflet.herokuapp.com/#super_cluster

Thank you, but one more question. Which version is needed for the supercluster function?, I am getting:
AttributeError: module 'dash_leaflet' has no attribute 'SuperCluster'

**Comment by emilhe** [comment]
I just implemented it, so you should use 0.0.22

**Comment by ELGUMOS1** [comment]
It works perflectly fine thanks @emilhe 
But is there a way to inyect the markers to the super_cluster via callback?
I got into the same situation as I told you in the previous days, this works pretty well when the markers are previously defined but if they are dynamically generated I still can not find the way to make them appear in the map and then be able to get their ID or any other info. This is the kind of pattern I am using.

`# Create example app.
app = dash.Dash()
app.layout = html.Div([
    dl.Map([
        dl.TileLayer(),
        dl.SuperCluster(data=[],id="sc"),
    ], center=(-37.8, 175.3), zoom=11, style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
    # Logging element.
    html.Button('Submit', id='submit-val', n_clicks=0)
])


@app.callback(Output("sc", "children"), [Input("submit-val", "n_clicks")])
def generate_markers(value):
    custom_markers = [
    dict(lat=-37.8, lon=174.95, tooltip="I am a tooltip"),  # marker with tooltip
    dict(lat=-37.8, lon=175.00, popup="I am a popup"),  # marker with popup
    dict(lat=-37.8, lon=175.05, markerOptions=dict(opacity=0.5)),  # marker with options
    dict(lat=-37.8, lon=175.10, ),  # custom icon
            ]
    return dlx.markers_to_geojson(custom_markers)

@app.callback(Output("log", "children"), [Input("sc", "featureClick"), Input("sc", "featureHover")])
def read_markers_info(click, hover):
    return json.dumps([click, hover])
    
if __name__ == '__main__':
    app.run_server()`

With one callback generate the markers based on the info from some dropdowns or buttons, then in another callback(if it is possible the same callback)  be able to read the markers id by clicking on them.  I have not found the correct pattern to achieve this behavior, any advice?
Thanks in advance, this proyect is amazing.

**Comment by emilhe** [comment]
@ELGUMOS1 Thanks! The `data` property of `SuperCluster` is static (at least for now), i.e. you can't modify it after a `SuperCluster` has been created. Instead, you should just create a new one, i.e. something along the lines of

```
import dash
import dash_html_components as html
import dash_leaflet as dl

from dash.dependencies import Input, Output

app = dash.Dash(prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.LayerGroup(id="container", children=[])], id="map", center=(56, 10), zoom=10,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
    html.Div(id="log")
])

@app.callback(Output("container", "children"), ....)
def update_super_cluster(...):
   data = # generate the data here
    return [dl.SuperCluster(data=data,id="sc")]

```

**Comment by ELGUMOS1** [comment]
Thanks a lot @emilhe I was very lost!
Whoever I tried that pattern several times and the markers in the map are still static, the data in the function changes dynamically based on the values from the dropdowns and buttons but the markers remain the same.
Any thoughts?

**Comment by emilhe** [comment]
@ELGUMOS1 Could you share a MWE? Otherwise it's that to guess what goes wrong. One common pitfall to mention though is that you should change the id of the layer (in this case the SuperCluster) to trigger the map to update. Here is a small example,

```
import dash
import dash_html_components as html
import dash_leaflet as dl
import dash_leaflet.express as dlx
import random

from dash.dependencies import Input, Output

app = dash.Dash(prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.LayerGroup(id="container", children=[])], id="map", center=(56, 10), zoom=10,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
    html.Button("Generate some markers", id="btn")
])


@app.callback(Output("container", "children"), [Input("btn", "n_clicks")])
def update_super_cluster(n_clicks):
    random_markers = [dict(lat=56 + 0.1 * random.random(), lon=10 + 0.1 * random.random()) for i in range(100)]
    return [dl.SuperCluster(data=dlx.markers_to_geojson(random_markers), id=str(n_clicks))]


if __name__ == '__main__':
    app.run_server()
```

**Comment by ELGUMOS1** [comment]
@emilhe  You are right! Now it works changing the id of the super cluster.
But now the issue is that I am not able to capture the featureClick event in a Callback because the super cluster id is not fixed and at the end my objective is to get the id of the marker by clicking on it.
Is there any way to achieve this behavior? I Just donated by the way, keep it up guys!

**Comment by emilhe** [comment]
Thanks! I guess you could achieve the desired behavior using pattern matching callbacks,

https://dash.plotly.com/pattern-matching-callbacks

I’ll make an example when I get to a PC. With your usecase I see that it would be simpler if the data property could be changed. I’ll look into that soon (:

**Comment by emilhe** [comment]
I have just created a (0.0.23rc1) release in which the `data` and `url` properties have been changed to dynamic. Hence, the following example should now work,

```
import json
import dash
import dash_html_components as html
import dash_leaflet as dl
import dash_leaflet.express as dlx
import random

from dash.dependencies import Input, Output

app = dash.Dash(prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map([dl.TileLayer(), dl.SuperCluster(id="sc")], center=(56, 10), zoom=10,
           style={'width': '100%', 'height': '50vh', 'margin': "auto", "display": "block"}),
    html.Button("Generate some markers", id="btn"),
    html.Div(id="log")
])


@app.callback(Output("sc", "data"), [Input("btn", "n_clicks")])
def update_super_cluster(n_clicks):
    random_markers = [dict(lat=56 + 0.1 * random.random(), lon=10 + 0.1 * random.random()) for i in range(100)]
    return dlx.markers_to_geojson(random_markers)


@app.callback(Output("log", "children"), [Input("sc", "featureClick")])
def super_cluster_click(feature):
    return json.dumps(feature)


if __name__ == '__main__':
    app.run_server()
```

**Comment by ELGUMOS1** [comment]
Thank so much @emilhe  It is exactly what I was looking for.
I finally developed the missing features. Incredible work.

**Comment by jbjuin** [comment]
@emilhe glad you improved on this ! Did you try with > 100k markers ?

On our side we went for a very low level & specific leaflet integration using PruneCluster (https://github.com/SINTEF-9012/PruneCluster) which flawlessly handle 300k markers.

Great work ! Congrats !

**Comment by emilhe** [comment]
Thanks! Yes, i tried with 1 million markers. It took ~ 2s to load, and after that the map was smooth :)

**Comment by sina7272** [comment]
Hi, Guys ,

I have two arrays which consists of lat and long and I want to created circles for each location same as below: 

for i in range (len(wellPositions)):

    lat = wellPositions[i][0]
    long = wellPositions[i][1]

markers = [dl.Marker(id=str(i), position=(lat,long) )]
cluster = dl.MarkerClusterGroup(children=markers, id="cluster")

app = dash.Dash(external_stylesheets=['https://codepen.io/chriddyp/pen/bWLwgP.css'], prevent_initial_callbacks=True)
app.layout = html.Div([
    dl.Map(children=[dl.TileLayer(url="https://a.tile.openstreetmap.org/{z}/{x}/{y}.png"), cluster],
           style={'width': "100%", 'height': "400px"}, center=[46, 2], zoom=7, id="map", preferCanvas=True), html.P(id="log")],
    style={"position": "relative", 'width': '1000px', 'height': '300px'})


@app.callback(Output("log", "children"), [Input("cluster", "click_marker_id")])
def marker_click(marker_id):
    return marker_id


I got no error but also I could not see any circle or point on my map. 

Can anyone help me please?

**Comment by ReactJsScc** [comment]
> Man, I really enjoy the UI of leaflets cluster map in your Dash wrapper- it's so nice for example how it consolidates/arranges many pins associated with the same point
> 
> ![image](https://user-images.githubusercontent.com/8431341/86999539-6b615a80-c178-11ea-9c5e-86d912e2eb9c.png)
> 
> if there is anyway to improve the performance to get 10s of thousands of these in cluster, I am all ears. Glad to help as possible too :)

brother if you didi this please tell me how can i do this i don't know much but i want to make like this round marker integration in map

**Comment by emilhe** [comment]
> > Man, I really enjoy the UI of leaflets cluster map in your Dash wrapper- it's so nice for example how it consolidates/arranges many pins associated with the same point
> 
> > 
> 
> > ![image](https://user-images.githubusercontent.com/8431341/86999539-6b615a80-c178-11ea-9c5e-86d912e2eb9c.png)
> 
> > 
> 
> > if there is anyway to improve the performance to get 10s of thousands of these in cluster, I am all ears. Glad to help as possible too :)
> 
> 
> 
> brother if you didi this please tell me how can i do this i don't know much but i want to make like this round marker integration in map
> 
> 

Could you post an MWE of you current code?
